### ISSUE TITLE
Using `undefined` or DEFAULT `authMechanism` fails

### ISSUE THREAD

**Issue by sommestad** [LearnBoost/mongoose#2619]
Following the 3.8.22 update (possibly the bump of the mongodb driver), we started getting connection issues (see below). 

*I'm unsure if this is actually a Mongoose bug, or if it's from the MongoDB driver but figured it's best to report it here, since this is the module we're using directly.*

It seems like the issues appear after in the following scenario: 
1. connect :+1: 
2. disconnect :+1: 
3. connect :bug: 

**Interim(?) solution**
Specifying an `authMechanism` in the connection options resolve the issue, even when using 3.8.22.

**Possible issue**
It seems like the issue appears around line 700 in `/mongodb/lib/mongodb/db.js`, where there's inconsistency in the if-statements. If the provided `authMechanism` is set to `DEFAULT`, it will cause an error.

Adding some logging shows that the `authMechanism` seems to be reused the next time `Db.prototype.authenticate()` is invoked, causing it to throw an error since it doesn't support an incoming value of `DEFAULT`.

**Notes:**
- We're using [a wrapping module](https://github.com/Springworks/node-mongoose-connection) to manage shared connections, but that hasn't changed for a long time. This module reuses existing connections to the same DB (across a top-level module and its submodules).
- Stepping down to 3.8.21 resolved the problem

**Error log**
```
{
    "timestamp": "2015-01-26T21:02:49.149Z",
    "name": "",
    "hostname": "",
    "pid": 42219,
    "level": 40,
    "err": {
        "name": "Error",
        "message": "only GSSAPI, PLAIN, MONGODB-X509, SCRAM-SHA-1 or MONGODB-CR is supported by authMechanism",
        "stack": "Error: only GSSAPI, PLAIN, MONGODB-X509, SCRAM-SHA-1 or MONGODB-CR is supported by authMechanism\n    at Db.authenticate (./node_modules/mongoose-connection/node_modules/mongoose/node_modules/mongodb/lib/mongodb/db.js:690:23)\n    at NativeConnection.Connection.onOpen (./node_modules/mongoose-connection/node_modules/mongoose/lib/connection.js:458:13)\n    at ./node_modules/mongoose-connection/node_modules/mongoose/lib/connection.js:421:10\n    at ./node_modules/mongoose-connection/node_modules/mongoose/lib/drivers/node-mongodb-native/connection.js:217:5\n    at ./node_modules/mongoose-connection/node_modules/mongoose/node_modules/mongodb/lib/mongodb/db.js:294:11\n    at process._tickDomainCallback (node.js:486:13)"
    },
    "msg": "only GSSAPI, PLAIN, MONGODB-X509, SCRAM-SHA-1 or MONGODB-CR is supported by authMechanism",
    "v": 0
}
```

**Comment by vkarpov15** [comment]
Opened a pull request for this in the driver https://github.com/mongodb/node-mongodb-native/pull/1246

Thanks for pointing this out!

**Comment by vkarpov15** [comment]
This was fixed for 4.0.0-rc2. I'll see about backporting this to 3.8.x for next release.
